src/libmw/include/libmw/defs.h

```diff
@@ -0,0 +1,238 @@
+#define DB_NAMESPACE namespace db {
+#define WALLET_NAMESPACE namespace wallet {
+#define END_NAMESPACE }
+
+// Forward Declarations
+namespace mw
+{
+    class Header;
+    class Block;
+    class BlockUndo;
+    class Transaction;
+    class ICoinsView;
+    class BlockBuilder;
+    struct State;
+}
+
+LIBMW_NAMESPACE
+
+typedef std::array<uint8_t, 32> BlockHash;
+typedef std::array<uint8_t, 32> KernelHash;
+typedef std::array<uint8_t, 32> Offset;
+typedef std::array<uint8_t, 32> BlindingFactor;
+typedef std::array<uint8_t, 33> Commitment;
+typedef std::array<uint8_t, 33> PubKey;
+typedef std::string MWEBAddress;
+
+static const uint8_t NORMAL_OUTPUT = 0;
+static const uint8_t PEGIN_OUTPUT = 1;
+
+/// <summary>
+/// Consensus parameters
+/// Any change to these will cause a hardfork!
+/// </summary>
+static constexpr size_t MAX_BLOCK_WEIGHT = 21'000;
+static constexpr size_t KERNEL_WEIGHT = 2;
+static constexpr size_t OWNER_SIG_WEIGHT = 1;
+static constexpr size_t OUTPUT_WEIGHT = 18;
+static constexpr uint16_t PEGIN_MATURITY = 20;
+static constexpr uint8_t MAX_KERNEL_EXTRADATA_SIZE = 33;
+
+/// <summary>
+/// Change outputs will use the stealth address generated using index 2,000,000.
+/// </summary>
+static constexpr uint32_t CHANGE_INDEX{ 2'000'000 };
+
+/// <summary>
+/// Peg-in outputs will use the stealth address generated using index 2,000,000.
+/// </summary>
+static constexpr uint32_t PEGIN_INDEX{ 4'000'000 };
+
+struct PegIn
+{
+    uint64_t amount;
+    std::array<uint8_t, 33> commitment;
+
+    bool operator==(const PegIn& rhs) const {
+        return amount == rhs.amount && commitment == rhs.commitment;
+    }
+
+    bool operator!=(const PegIn& rhs) const {
+        return !(*this == rhs);
+    }
+};
+
+struct PegOut
+{
+    uint64_t amount;
+    std::vector<uint8_t> scriptPubKey;
+};
+
+struct HeaderRef
+{
+    std::shared_ptr<const mw::Header> pHeader;
+};
+
+struct HeaderAndPegsRef
+{
+    HeaderRef header;
+    std::vector<PegIn> pegins;
+    std::vector<PegOut> pegouts;
+};
+
+/// <summary>
+/// A simple interface for accessing members of an MWEB block.
+/// </summary>
+struct BlockRef
+{
+    /// <summary>
+    /// Checks whether the internal block pointer is null.
+    /// If it's null, it is unsafe to call any other methods on this object.
+    /// </summary>
+    /// <returns>true if interal pointer is null. Otherwise, false.</returns>
+    bool IsNull() const noexcept { return pBlock == nullptr; }
+
+    MWIMPORT libmw::BlockHash GetHash() const noexcept;
+    MWIMPORT libmw::HeaderRef GetHeader() const noexcept;
+    MWIMPORT uint64_t GetTotalFee() const noexcept;
+    MWIMPORT uint64_t GetWeight() const noexcept;
+    MWIMPORT std::set<KernelHash> GetKernelHashes() const noexcept;
+    MWIMPORT std::vector<libmw::Commitment> GetInputCommits() const noexcept;
+    MWIMPORT std::vector<libmw::Commitment> GetOutputCommits() const noexcept;
+
+    std::shared_ptr<mw::Block> pBlock;
```

Comment:
const mw::Block?

Reply:
A few places require a mutable block (to mark it as validated).